### PR TITLE
Migration api now works: first draft

### PR DESCRIPTION
**Pull request by juandent**
connection_holder() can now be copied. DbConnection holds a connection through any number of make_storage() calls.
There does  not seem to be a problem with retain counts because all objects are destructed and the connection closed at DbConnection destructor.
In any case the method needs inspection to be sure retain counts are not less or more than they should:

```C++
connection_holder(const connection_holder& rhs) :
                filename{rhs.filename}, db(rhs.db), _retain_count(rhs._retain_count) {
                retain(); 
            }
``` 

Also look at storage_base() protected constructor with the DbConnection overload!

### CODE REVIEW

**Comment by juandent** [comment]
> `connection_holder` is not designed to be copied. Every connection (`sqlite3*`) must be unique. Yes it is a pointer and can be copied. That is why `connection_holder` is designed to be non copyable. Cause if everyone will be able to copy connections then you'll never know where all same connections are stored

ok
So you have a better idea for DbConnection class?

**Comment by juandent** [comment]
but if we take retain count into account? it would be like implementing a std::shared_ptr<>
?
what is the problem with that?

> Every connection (sqlite3*) must be unique
why?

**Comment by juandent** [comment]
Ok here is what I see:

DbConnection opens a sqlite3* like this:

```C++
        DbConnection(const std::string& filename) :
            connection(std::make_unique<internal::connection_holder>(filename)) {
            this->connection->retain();
            db = this->connection->get();
        }
```

since connection in storage_base is ` std::unique_ptr<connection_holder> connection;` and it is initialized in the storage_base constructor with a filename accepting constructor. This instance of connection_holder initializes sqlite3* to nullptr and _retain_count to 0. Therefore when a connection is required, retain() is called on the connection_holder which opens the database using the filename parameter like this:

```C++
            void retain() {
                ++this->_retain_count;
                if(1 == this->_retain_count) {
                    auto rc = sqlite3_open(this->filename.c_str(), &this->db);
                    if(rc != SQLITE_OK) {
                        throw_translated_sqlite_error(db);
                    }
                }
            }
```
Therefore a new db pointer is associated with this storage and not with the db pointer in DbConnection... How then to force storage_base to use DbConnection's db pointer? They would have to share the connection_holder!


I see no other solution. I will appreciate comments on this points above...

What damage comes from sqlite3* being the same for every storage? Isn't that what we want with DbConnection?

**Comment by juandent** [comment]
>  Every connection (sqlite3*) must be unique. 

every storage has its own connection_holder which opens the database and obtains a particular sqlite3 pointer different from the other storages!! How do we reconcile this?

**Comment by juandent** [comment]
In view of my arguments above, why not change the connection_holder to shared_ptr<>?

this way all storages can SHARE a connection_holder which is created in DbConnection class??
For the rest of storage_base it will be transparent - should not be affected at all! they continue having their own connection_holder associated with a filename

RECALL: 

1. we want storages to share sqlite3*, 
2. currently there is a 1:1 relation between connection_holder and sqlite3*
3. therefore if we want to share sqlite3* => we need to share connectioon_holder

**Comment by juandent** [comment]
Notice that connection_holder isn't copy constructible as you wished:

```C++
connection_holder(const connection_holder&) = delete;
```

**Comment by juandent** [comment]
To test use this:

```C++
#include "Schema_v1.h"

using namespace sqlite_orm;

using namespace std::literals;

using migrator = std::function<void(const DbConnection&)>;

std::map<std::pair<size_t, size_t>, migrator> migrations; // <version pair, lambda>

void register_migration(size_t from, size_t to, migrator m)
{
    auto p = std::make_pair(from, to);
    migrations[p] = m;
}

const std::string dbFileName{ "migration_exercise.sqlite"s };

void migrate_to(int target_version, const DbConnection& con) {
    int current{ 0 };
    {
        auto storage = make_storage(con);    // just enough to get the version!
        current = storage.pragma.user_version();
    }
    auto p = std::make_pair(current, target_version);
    if (migrations.find(p) != migrations.end())
    {
        auto lambda = migrations[p];
        lambda(con);
        return;
    }

    while(current < target_version) {
        auto p = std::make_pair(current, current+1);
        if(migrations.find(p) != migrations.end()) {
            auto lambda = migrations[p];
            lambda(con);
            ++current;
        }
        else
            break;
    }
}

// dummy database
void prepareDbForVersion1() {
    // insert using the original schema!
    struct OldUser {
        int id = 0;
        std::string name;
    };
    auto oldStorage = make_storage(dbFileName,
        make_table("users",
            make_column("id", &OldUser::id, primary_key()),
            make_column("name", &OldUser::name)));
    oldStorage.sync_schema();
    oldStorage.remove_all<OldUser>();
    oldStorage.pragma.user_version(1);
    oldStorage.replace(OldUser{ 1, "juan" });
    oldStorage.replace(OldUser{ 2, "klaus" });
    oldStorage.replace(OldUser{ 3, "eugene" });
}
``` 

```C++
        prepareDbForVersion1();
        DbConnection con{ dbFileName };

        register_migration(1, 2, [](const DbConnection& con) {
            std::map<int, std::string> emails = {
                { 1, "fnc12@me.com" },
                { 2, "megasuperhero@gmail.com" },
                { 3, "thor@hotmail.com" },
            };
            auto oldStorage = version1::getStorage(con);
            auto allOldUsers = oldStorage.get_all<version1::User>();
            auto nextStorage = version2::getStorage(con);

            nextStorage.sync_schema(); // old users are dropped here
            nextStorage.pragma.user_version(2);

            for (auto& oldUser : allOldUsers) {
                version2::User newUser{ oldUser.id, move(oldUser.name), emails[oldUser.id] };
                nextStorage.insert(newUser);
            }
            });


        register_migration(2, 3, [](const DbConnection& con) {
            auto oldStorage = version2::getStorage(con);
            auto allOldUsers = oldStorage.get_all<version2::User>();
            auto nextStorage = version3::getStorage(con);

            nextStorage.sync_schema(); // old users are dropped here
            nextStorage.pragma.user_version(3);

            for (auto& oldUser : allOldUsers) {
                auto& oldUserName = oldUser.name;
                auto spaceIndex = oldUserName.find(' ');
                version3::User newUser{ oldUser.id };
                if (spaceIndex != oldUserName.npos) { // space is found
                    auto firstName = oldUserName.substr(0, spaceIndex);
                    auto lastName = oldUserName.substr(spaceIndex, oldUserName.size() -
                        spaceIndex);
                    newUser.first_name = move(firstName);
                    newUser.last_name = move(lastName);
                }
                else {
                    newUser.first_name = move(oldUserName);
                    newUser.last_name = {};
                }
                nextStorage.insert(newUser);
            }
        });
        migrate_to(3, con);
```
I will attach a file

**Comment by juandent** [comment]
Schema_v1.h contains:


```C++
#include <string>
#include <sqlite_orm/sqlite_orm.h>

namespace version1 {

    using namespace sqlite_orm;

    struct User {
        int id = 0;
        std::string name;
    };

    auto getStorage(const DbConnection& con)
    {
        auto storage = make_storage(con,
            make_table("users",
                make_column("id", &User::id, primary_key()),
                make_column("name", &User::name)));
        return storage;
    }
}

namespace version2 {
    using namespace sqlite_orm;

    struct User {
        int id = 0;
        std::string name;
        std::string email;
    };

    auto getStorage(const DbConnection& con)
    {
        auto storage = make_storage(con,
            make_table("users",
                make_column("id", &User::id, primary_key()),
                make_column("name", &User::name),
                make_column("email", &User::email)));
        return storage;
    }
}

namespace version3 {
    using namespace sqlite_orm;

    struct User {
        int id = 0;
        std::string first_name;
        std::string last_name;
        std::string email;
    };

    auto getStorage(const DbConnection& con)
    {
        auto storage = make_storage(con,
            make_table("users",
                make_column("id", &User::id, primary_key()),
                make_column("first_name", &User::first_name),
                make_column("last_name", &User::last_name),
                make_column("email", &User::email)));
        return storage;
    }
}
```

**Comment by juandent** [comment]
you can now see DbConnection in action in the test file MigrationAPI.cpp
No need to copy paste bunch of lines I copied above, better run MigrationAPI.cpp using debugger
It uses a trick by setting an storage's on_open member to catch if any calls to that storage opens a new connection: it doesn't, so it is reusing the connection in DbConnection object used to open the storage

**Comment by juandent** [comment]
All migration API seems implemented! Please review!

**Comment by fnc12** [comment]
> So you have a better idea for DbConnection class?

`using DbConnection = sqlite3*;` just for now. Later we can refactor it to a container class which contains `sqlite3*`.

> but if we take retain count into account? it would be like implementing a std::shared_ptr<>

`connection_ref` is designed to be a smart pointer to `connection_holder` but you reimplementing it removing `connection_holder`s design goal. It is not good. If you want to keep a reference to a connection use `connection_ref` or raw `sqlite3*` just for now.

> What damage comes from sqlite3* being the same for every storage? Isn't that what we want with DbConnection?

it is ok and it is how it must be. But without copying `connection_holder`  cause holder is designed to be unique. Otherwise two connection holders with the same connections will call `sqlite3_close` for the same argument in dtors and it is wrong. We need to think how to do that correctly.

> therefore if we want to share sqlite3* => we need to share connection_holder

yes it sounds like a solution but shared pointer has one important disadvantage: you are not sure whether dropping a shared pointer will destroy the object. This is why `connection_holder` was created: it acts like shared pointer storage with `connection_ref`s as weak pointers to it and you always know where actual data is stored. So I'd better change `storage_base` to keep a `connection_ref` instead of `connection_holder` in cases when we create a storage without a filename. What you say is right. Better to implement it with keeping `connection_holder` as a unique holder.

Ok I read all your comments. Let me make a review after my breakfast. Thanks Juan

**Comment by juandent** [comment]
>But without copying connection_holder cause holder is designed to be unique. Otherwise two connection holders with the same connections will call sqlite3_close for the same argument in dtors and it is wrong. We need to think how to do that correctly.

Isn't this taken care by reference counting?

**Comment by juandent** [comment]
>copying connection_holder

But now we are sharing connection_holder so they might have access to the same sqlite3 pointer! Don't see the problem -- it should be as simple as making sure the reference counts go up and down correctly which in turn assures that db will not be closed twice!

**Comment by juandent** [comment]
Making the connection member of storage_base a shared_ptr<connection_holder> correctly opens the database when DbConnection class is constructed and closes the database, upon DbConnection destructor with retain/release working perfectly across several storages initialized with the DbConnection instance!

**Comment by juandent** [comment]
The one job that is failing is with VS 2015 update 3, x64

File: tests/MigrationAPI.cpp [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    oldStorage.replace(OldUser{2, "klaus"});
+    oldStorage.replace(OldUser{3, "eugene"});
+    retain_count = oldStorage.get_connection_holder()->retain_count();
+    REQUIRE(retain_count == 1);
+    REQUIRE(con.connection.use_count() == 2);
+    REQUIRE(pDb == nullptr);
+
+    SECTION("register_migrations") {
+        oldStorage.register_migration(1, 2, [](const DbConnection& con) {
+            internal::connection_ref{*con.connection};
```
Comment (fnc12) [inline tests/MigrationAPI.cpp:117]: redundant line

File: tests/MigrationAPI.cpp [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+                {3, "thor@hotmail.com"},
+            };
+            auto oldStorage = version1::getStorage(con);
+            auto retain_count = oldStorage.get_connection_holder()->retain_count();
+            REQUIRE(retain_count == 1);
+            auto allOldUsers = oldStorage.get_all<version1::User>();
+            auto nextStorage = version2::getStorage(con);
+            retain_count = nextStorage.get_connection_holder()->retain_count();
+            nextStorage.sync_schema();  // old users are dropped here
+            nextStorage.pragma.user_version(2);
```
Comment (fnc12) [inline tests/MigrationAPI.cpp:130]: don't set `user_version` manually cause storage can do it by itself inside `migrate_to` call

File: tests/MigrationAPI.cpp [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            retain_count = nextStorage.get_connection_holder()->retain_count();
+            REQUIRE(retain_count == 1);
+            for(auto& oldUser: allOldUsers) {
+                version2::User newUser{oldUser.id, move(oldUser.name), emails[oldUser.id]};
+                nextStorage.insert(newUser);
+            }
+            auto count = con.connection.use_count();
```
Comment (fnc12) [inline tests/MigrationAPI.cpp:137]: redundant line

File: tests/MigrationAPI.cpp [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+            nextStorage.sync_schema();  // old users are dropped here
+            nextStorage.pragma.user_version(3);
+
+            for(auto& oldUser: allOldUsers) {
+                auto& oldUserName = oldUser.name;
+                auto spaceIndex = oldUserName.find(' ');
+                version3::User newUser{oldUser.id};
+                if(spaceIndex != oldUserName.npos) {  // space is found
+                    auto firstName = oldUserName.substr(0, spaceIndex++);
```
Comment (fnc12) [inline tests/MigrationAPI.cpp:153]: I bet `spaceIndex++` must be replaced with `spaceIndex++ - 1`

File: tests/MigrationAPI.cpp [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+                    auto lastName = oldUserName.substr(spaceIndex, oldUserName.size() - spaceIndex);
```
Comment (fnc12) [inline tests/MigrationAPI.cpp:154]: `auto lastName = oldUserName.substr(spaceIndex + 1, oldUserName.size() - spaceIndex - 1);`

File: tests/MigrationAPI.cpp [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+                } else {
+                    newUser.first_name = move(oldUserName);
+                    newUser.last_name = {};
+                }
+                newUser.email = oldUser.email;
+                nextStorage.insert(newUser);
+            }
+        });
+
+        oldStorage.migrate_to(3, con);
```
Comment (fnc12) [inline tests/MigrationAPI.cpp:166]: no need to pass `con` cause storage already has one

File: tests/MigrationAPI.cpp [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+        REQUIRE(1 == 1);
```
Comment (fnc12) [inline tests/MigrationAPI.cpp:167]: redundant line

**Review by fnc12 — CHANGES_REQUESTED** [review]

**Review by fnc12 — CHANGES_REQUESTED** [review]

File: dev/DbConnection.h [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+namespace sqlite_orm {
+
+    struct DbConnection {
+        DbConnection(const std::string& filename) :
+            connection{std::make_shared<internal::connection_holder>(filename)} {
+            this->connection->retain();  // opens the connection and stores the sqlite3*
+        }
+
+        DbConnection(const DbConnection& o) {
+            this->connection = o.connection;
```
[the remark this answers is not in the document]
Comment (juandent) [reply dev/DbConnection.h:14]: no we need to copy the shared_ptr<> which is connection... recall it is a shared_ptr<> now!! not a unique_ptr<>!!

File: dev/DbConnection.h [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+            this->connection->retain();
+        }
+
+        ~DbConnection() {
+            this->connection->release();
+        }
+        std::string get_filename() const {  // so we can access filename from make_storage(DbConnection,...)
+            return this->connection->filename;
+        }
+        std::shared_ptr<internal::connection_holder> connection;
```
[the remark this answers is not in the document]
Comment (juandent) [reply dev/DbConnection.h:24]: make the connection member private?? We need it public for the storage_base() constructor!!
look at this:

```C++
storage_base(const DbConnection& con, int foreignKeysCount) :
                pragma(std::bind(&storage_base::get_connection, this)),
                limit(std::bind(&storage_base::get_connection, this)),
                inMemory(con.get_filename().empty() || con.get_filename() == ":memory:"), connection(con.connection),
                cachedForeignKeysCount(foreignKeysCount), initByConnection{true} {}
``` 

You see why it needs to be public?

File: tests/MigrationAPI.cpp (hunk 4 above)
Comment (juandent) [reply tests/MigrationAPI.cpp:153]: no not really

File: dev/storage_base.h [hunk 10]
```diff
@@ -190,6 +192,31 @@ namespace sqlite_orm {
                 }
             }
 
+            void register_migration(size_t from, size_t to, migrator m) {
+                auto p = std::make_pair(from, to);
+                this->migrations[p] = m;
```
[the remark this answers is not in the document]
Comment (juandent) [reply dev/storage_base.h:197]: ok

File: dev/storage_base.h [hunk 11]
```diff
@@ -190,6 +192,31 @@ namespace sqlite_orm {
...
+            }
+
+            void migrate_to(int target_version, const DbConnection& con) {
+                int from_version = pragma.user_version();
```
[the remark this answers is not in the document]
Comment (juandent) [reply dev/storage_base.h:201]: ok

File: dev/storage_base.h [hunk 12]
```diff
@@ -190,6 +192,31 @@ namespace sqlite_orm {
...
+                auto p = std::make_pair(from_version, target_version);
+                if(migrations.find(p) != migrations.end()) {
+                    auto lambda = migrations[p];
```
[the remark this answers is not in the document]
Comment (juandent) [reply dev/storage_base.h:204]: ok

File: dev/storage_base.h [hunk 13]
```diff
@@ -190,6 +192,31 @@ namespace sqlite_orm {
...
+                    return;
+                }
+
+                while(from_version < target_version) {
+                    auto p = std::make_pair(from_version, from_version + 1);
+                    if(migrations.find(p) != migrations.end()) {
+                        auto lambda = migrations[p];
+                        lambda(con);
+                        ++from_version;
+                    } else
```
[the remark this answers is not in the document]
Comment (juandent) [reply dev/storage_base.h:215]: ok

File: dev/storage_base.h [hunk 14]
```diff
@@ -190,6 +192,31 @@ namespace sqlite_orm {
...
+                if(migrations.find(p) != migrations.end()) {
+                    auto lambda = migrations[p];
+                    lambda(con);
+                    return;
+                }
+
+                while(from_version < target_version) {
+                    auto p = std::make_pair(from_version, from_version + 1);
+                    if(migrations.find(p) != migrations.end()) {
+                        auto lambda = migrations[p];
```
[the remark this answers is not in the document]
Comment (juandent) [reply dev/storage_base.h:212]: correct

File: dev/storage_base.h [hunk 15]
```diff
@@ -190,6 +192,31 @@ namespace sqlite_orm {
...
+                        lambda(con);
+                        ++from_version;
+                    } else
+                        break;
```
[the remark this answers is not in the document]
Comment (juandent) [reply dev/storage_base.h:216]: ok

File: dev/storage_base.h [hunk 16]
```diff
@@ -487,12 +514,24 @@ namespace sqlite_orm {
                 }
             }
 
+            connection_holder* get_connection_holder() const {
+                return this->connection.get();
+            }
+
           protected:
+            // jdh
```
[the remark this answers is not in the document]
Comment (juandent) [reply dev/storage_base.h:522]: ok removed

File: tests/MigrationAPI.cpp [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+        std::string name;
+    };
+}
+
+TEST_CASE("DbConnectionAPI") {
+    // insert using the original schema!
+    DbConnection con{"migration_api.sqlite"};
+    REQUIRE(con.connection.use_count() == 1);
+
+    auto oldStorage = make_storage(
```
[the remark this answers is not in the document]
Comment (juandent) [reply tests/MigrationAPI.cpp:91]: you mean it is created inside migrate_to(3)? 
```C++
pass sqlite3* between two storages within a migration
``` 

that is why connection is:

```C++
 std::shared_ptr<internal::connection_holder> connection;
```

[the remark this answers is not in the document]
Comment (juandent) [reply tests/MigrationAPI.cpp:91]: so it can SHARE the connection_holder!

**Pull request closed without merging**